DataFormats/Detectors/TPC/include/DataFormatsTPC/PIDResponse.h

```diff
@@ -0,0 +1,117 @@
+
+// o2 includes
+#include "GPUCommonDef.h"
+#include "GPUCommonRtypes.h"
+#include "GPUCommonMath.h"
+#include "ReconstructionDataFormats/PID.h"
+#include "DataFormatsTPC/PIDResponse.h"
+#include "DataFormatsTPC/BetheBlochAleph.h"
+#include "DataFormatsTPC/TrackTPC.h"
+
+namespace o2::tpc
+{
+class TrackTPC;
+
+/// \brief PID response class
+///
+/// This class is used to handle the TPC PID response.
+///
+
+class PIDResponse
+{
+ public:
+  /// default constructor
+  PIDResponse() CON_DEFAULT;
+
+  /// default destructor
+  ~PIDResponse() CON_DEFAULT;
+
+  /// setters
+  GPUd() void setBetheBlochParams(const float betheBlochParams[5]);
+  GPUd() void setMIP(float mip) { mMIP = mip; }
+  GPUd() void setChargeFactor(float chargeFactor) { mChargeFactor = chargeFactor; }
+
+  /// getters
+  GPUd() const float* getBetheBlochParams() const { return mBetheBlochParams; }
+  GPUd() float getMIP() const { return mMIP; }
+  GPUd() float getChargeFactor() const { return mChargeFactor; }
+
+  /// get expected signal of the track
+  GPUd() float getExpectedSignal(const TrackTPC& track, const o2::track::PID::ID id) const;
+
+  /// get most probable PID of the track
+  GPUd() o2::track::PID::ID getMostProbablePID(const TrackTPC& track) const;
+
+ private:
+  float mBetheBlochParams[5] = {0.19310481, 4.26696118, 0.00522579, 2.38124907, 0.98055396}; // BBAleph average fit parameters
+  float mMIP = 50.f;
+  float mChargeFactor = 2.299999952316284f;
+
+#ifndef GPUCA_ALIROOT_LIB
+  ClassDefNV(PIDResponse, 1);
+#endif
+};
+
+GPUd() void PIDResponse::setBetheBlochParams(const float betheBlochParams[5])
+{
+  for (int i = 0; i < 5; i++) {
+    mBetheBlochParams[i] = betheBlochParams[i];
+  }
+}
+
+GPUd() float PIDResponse::getExpectedSignal(const TrackTPC& track, const o2::track::PID::ID id) const
+{
+  const float bg = static_cast<float>(track.getP() / o2::track::pid_constants::sMasses[id]);
+  if (bg < 0.05) {
+    return -999.;
+  }
+  const float bethe = mMIP * o2::tpc::BetheBlochAleph(bg, mBetheBlochParams[0], mBetheBlochParams[1], mBetheBlochParams[2], mBetheBlochParams[3], mBetheBlochParams[4]) * o2::gpu::GPUCommonMath::Pow(static_cast<float>(o2::track::pid_constants::sCharges[id]), mChargeFactor);
+  return bethe >= 0. ? bethe : -999.;
+}
+
+// get most probable PID
+GPUd() o2::track::PID::ID PIDResponse::getMostProbablePID(const TrackTPC& track) const
+{
+  const float dEdx = track.getdEdx().dEdxTotTPC;
+
+  if (dEdx < 10) {
+    return o2::track::PID::Pion;
+  }
+
+  auto id = o2::track::PID::Electron;
+  float distanceMin = o2::gpu::GPUCommonMath::Abs(dEdx - getExpectedSignal(track, id));
+
+  // calculate the distance to the expected dEdx signals
+  // start from Pion to exlude Muons
+  for (o2::track::PID::ID i = o2::track::PID::Pion; i < o2::track::PID::NIDs; i++) {
+    const float distance = o2::gpu::GPUCommonMath::Abs(dEdx - getExpectedSignal(track, i));
+    if (distance < distanceMin) {
+      id = i;
+      distanceMin = distance;
+    }
+  }
+
```

Comment:
@tubagundem @wiechula : In the aliroot we were overriding the electrons to the closest alternative PID in the bands crossing regions: https://github.com/alisw/AliRoot/blob/f01cd3609506f29aa9b5f0c408b4234c99325a61/STEER/ESD/AliESDpid.cxx#L472-L480. I think we should introduce the same here, perhaps passing the bands and tolerances definitions via configurable params.

Reply:
If I understood correctly and add some code like below, you would like to pass these `(p > 0.38) && (p < 0.48)`  and `(p > 0.75) && (p < 0.85)` via configurable params and not hardcoded right?

 `// override the electrons to the closest alternative PID in the bands crossing regions
  if (id == o2::track::PID::Electron && useElectronExclusionBands) {
    const float p = track.getP();
    if ((p > 0.38) && (p < 0.48)) {
        id = o2::track::PID::Kaon;
    } else if ((p > 0.75) && (p < 0.85)) {
        id = o2::track::PID::Proton;
    }
  }`
  
Then do I need to add another option to reco workflow below?
[](https://github.com/AliceO2Group/AliceO2/blob/1af9293cdfd6c8f7005998c7ea2dce3dc7d70dc6/Detectors/TPC/workflow/src/tpc-reco-workflow.cxx#L57)

Reply:
Yes, but I would do this via configurable params, e.g. adding after https://github.com/AliceO2Group/AliceO2/blob/1af9293cdfd6c8f7005998c7ea2dce3dc7d70dc6/GPU/GPUTracking/Definitions/GPUSettingsList.h#L58

```
AddOptionRTC(PID_EKrangeMin, float, 0.38, "", 0, "min P of electron/K BB bands crossing")
AddOptionRTC(PID_EKrangeMax, float, 0.48, "", 0, "max P of electron/K BB bands crossing")
AddOptionRTC(PID_EPrangeMin, float, 0.75, "", 0, "min P of electron/p BB bands crossing")
AddOptionRTC(PID_EPrangeMax, float, 0.85, "", 0, "max P of electron/p BB bands crossing")
```
unless @davidrohr may suggest a more appropriate place.
The numeric values later should be tuned for run3.

Reply:
> unless @davidrohr may suggest a more appropriate place.
> The numeric values later should be tuned for run3.

fine wth me

Reply:
I reconstructed one ctf file to test it. It looks like below before and after I added the overriding of electrons.

![image](https://github.com/AliceO2Group/AliceO2/assets/48834043/6a6459a9-9a92-4a8a-b062-da47ccd47afc)

Reply:
but why we don't see the bands?

Reply:
It has a very slight effect

![image](https://github.com/AliceO2Group/AliceO2/assets/48834043/5df00eea-f085-4ba3-ad0d-131e3183787c)

Reply:
Probably the markers are too large so that the density is visible except on the periphery. Anyway, judging about the crossing bands on the log plot of this scale makes no sense, they should contain ~10% of K and p, which could be lost under the electron hypothesis.